### PR TITLE
Eliminate nearly all Qt widget leaks by using qtbot

### PR DESCRIPTION
**Pull request by Czaki**
# Description

This PR is a next PR to clean up bugs in tests that are causing segfaults for some test run. Previous one was #5373

In this PR I have focused on finding widgets that are not properly teardown. 

Next to the improving of tests:

1) change part of the test to use `QtViewer` instead of full `Viewer` 
2) add missed widget to qtbot 
3) use `before_close_func` instead of custom teardown code in fixture

This PR contains a code for detecting of leaked widgets. This code contains two parts. 

The own subclass of QtBot that ensures that object, with is hadled using `addWidget` name is set to `'handled_widget'` affter the test end. 
And `_find_dangling_widgets` fixture that is executed after the test end (and after qtbot teardown) and looks for all top-level widgets that do not have parents and check if this widgets are either:

1. `_QtMainWindow` - we handle it using `make_napari_viewer` fixture, without using qtbot. 
2. some widget form qtconsole (I do not have time to investigate if we could make a PR to qtconsole to avoid it).
3. is not `CanvasBackendDesktop` that is leaking in `test_sys_info.py` and may require PR to vispy to fix this. (or some monkeypatching)
4. is not handled by qtbot already by having name `"handled_widget"` 
5. add fixtures for `viewer_model` and `qt_viewer` 

If the widget does not fit to any above category, it is reported as an error. 

There is no guarantee that all problematic widgets are found. So if this PR will be merged then some random test failures need to be observed, as some widget not too often leak from tests.

### CODE REVIEW

File: napari/_qt/qt_viewer.py [hunk 1]
```diff
@@ -1174,7 +1174,8 @@ def closeEvent(self, event):
         event : qtpy.QtCore.QCloseEvent
             Event from the Qt context.
         """
-        self.layers.close()
+        if self._layers is not None:
+            self.layers.close()
```
Comment (Czaki) [inline napari/_qt/qt_viewer.py]:
```suggestion
        if self._layers is not None:
        	# do not create layerlist if it do not exist yet. 
            self.layers.close()
```

File: napari/_qt/_tests/test_qt_notifications.py [hunk 2]
```diff
@@ -49,6 +49,7 @@ def _clean_current(monkeypatch, qtbot):
     base_show = NapariQtNotification.show
 
     widget = QWidget()
+    # widget.setObjectName('welcome_widget')
```
Comment (Czaki) [inline napari/_qt/_tests/test_qt_notifications.py]:
```suggestion
```

File: napari/_qt/_tests/test_qt_notifications.py [hunk 3]
```diff
@@ -147,8 +148,8 @@ def test_notification_manager_via_gui(
     Test that the notification_manager intercepts `sys.excepthook`` and
     `threading.excepthook`.
     """
-    errButton = QPushButton()
-    warnButton = QPushButton()
+    errButton = QPushButton('err')
+    warnButton = QPushButton('warn')
```
Comment (Czaki) [inline napari/_qt/_tests/test_qt_notifications.py]:
```suggestion
    errButton = QPushButton()
    warnButton = QPushButton()
```

**Comment by lucyleeow** [comment]
Thank you @Czaki, this is great work! I am not knowledgable enough in Qt to review but there is some very useful information here that would be great to get into our docs (probably in [testing](https://napari.org/dev/developers/contributing/testing.html#testing-with-qt-and-napari-viewer)). I can think of these items:

* advise when to use `make_napari_viewer` and when `QtViewer` is adequate
* advise on the autouse fixtures we add when `qapp` is used (I had no idea of this)
* how `_find_dangling_widgets` checks for widgets (essentially your bullet list above, it's very useful)

**Comment by Czaki** [comment]
> advise when to use `make_napari_viewer` and when `QtViewer` is adequate

I have put this in "Use the smallest possible widgets for tests" section of https://czaki.github.io/blog/2024/09/16/preventing-segfaults-in-test-suite-that-has-qt-tests/ feel free to take information from there to docs. 

> advise on the autouse fixtures we add when `qapp` is used (I had no idea of this)

I do not want to advertise it too much. It is a kind of magic, that may introduce a confusion to contributors. I much more prefer split on Qt and non Qt tests based on simple directory structure and using `@pytest.fixture(autouse=True)` in non-top level `conftest.py`. But in napari it will require to keep tests outside the code. That means a big refactor. 

> how `_find_dangling_widgets` checks for widgets (essentially your bullet list above, it's very useful)

I would rather put this in a comment in code. It may outdated easy, and if it is documented in code (like `_dangling_qthreads`) it is much easier to spot this during code review. 

However, experience from writing `_dangling_qthreads` shows that my initial comments were totally rewritten by @jni to be clear by less experienced user.

**Comment by lucyleeow** [comment]
> I have put this in "Use the smallest possible widgets for tests" section of https://czaki.github.io/blog/2024/09/16/preventing-segfaults-in-test-suite-that-has-qt-tests/ feel free to take information from there to docs.

I mean specifically for napari case. For dummies, can you say what does `Viewer` have that `QtViewer` does not (and thus in what case you would need to use full `Viewer`) ? I'm happy to add a section on this if you give me a rough guide?

> I would rather put this in a comment in code.

Sure, would be great to see this in the code.

**Comment by Czaki** [comment]
> I mean specifically for napari case. For dummies, can you say what does `Viewer` have that `QtViewer` does not (and thus in what case you would need to use full `Viewer`) ? I'm happy to add a section on this if you give me a rough guide?

Based on my knowledge, the main difference is the option to dock widgets (especially all plugin widgets), and has menus. 
It is used, also, to detect active window (but it may be a problem in tests).

Fast checking of code does not show more usecases. 

Using a pair of `viewer = ViewerModel(); qt_viewer = QtViewer(viewer)` makes also `viewer.window` path unavailable, so do not allow testing this interface.

File: napari/_qt/_tests/test_qt_viewer.py [hunk 4]
```diff
@@ -744,16 +744,21 @@ def _update_data(
...
     qt_viewer.show()
     qt_viewer.controls.show()
-    yield qt_viewer
-    qt_viewer.controls.hide()
-    qt_viewer.controls.close()
-    qt_viewer.hide()
-    qt_viewer.close()
-    qt_viewer._instances.clear()
-    qtbot.wait(50)
+    return qt_viewer
```
Comment (jni) [inline napari/_qt/_tests/test_qt_viewer.py:748]: These changes are super interesting! Are they equivalent except that instead of waiting a specified period of time, qtbot actually makes sure that the controls are closed?

File: napari/_qt/layer_controls/qt_image_controls_base.py [hunk 5]
```diff
@@ -152,6 +152,9 @@ def __init__(self, layer: Image) -> None:
         self.colorbarLabel.setToolTip(trans._('Colorbar'))
 
         self._on_colormap_change()
+        if self.__class__ == QtBaseImageControls:
+            # this if is to prevent leaking of buttons in the tests
+            self.layout().addRow(self.button_grid)
```
Comment (jni) [inline napari/_qt/layer_controls/qt_image_controls_base.py]: I don't understand this at all... Could you comment on what was happening before, and how this change fixes it?

File: napari/_qt/_tests/test_qt_viewer.py (hunk 4 above)
Comment (Czaki) [reply napari/_qt/_tests/test_qt_viewer.py:748]: Yes. As I modified `make_napari_viewer` to not raise exception on `'handled_widget'` as leaked widget, we do not need to wait until every data is cleaned. We could allow `qtbot` to handle it in a common way.

File: napari/_qt/layer_controls/qt_layer_controls_base.py [hunk 6]
```diff
@@ -157,6 +157,9 @@ def __init__(self, layer: Layer) -> None:
         self.opacityLabel.setEnabled(
             self.layer.blending not in NO_OPACITY_BLENDING_MODES
         )
+        if self.__class__ == QtLayerControls:
+            # this if is to prevent leaking of buttons in the tests
+            self.layout().addRow(self.button_grid)
```
Comment (jni) [inline napari/_qt/layer_controls/qt_layer_controls_base.py:165]: As above — can you explain why things were leaking, and how this fixes that issue?

File: napari/_qt/qt_viewer.py [hunk 7]
```diff
@@ -1174,7 +1174,9 @@ def closeEvent(self, event):
         event : qtpy.QtCore.QCloseEvent
             Event from the Qt context.
         """
-        self.layers.close()
+        if self._layers is not None:
+            # do not create layerlist if it do not exist yet.
```
Comment (jni) [inline napari/_qt/qt_viewer.py]:
```suggestion
            # do not create layerlist if it does not exist yet.
```

File: napari/_vispy/_tests/test_vispy_labels_layer.py [hunk 8]
```diff
@@ -68,21 +72,21 @@ def test_labels_fill_slice(make_napari_viewer, array_type):
     layer.n_edit_dimensions = 3
     QCoreApplication.instance().processEvents()
     layer.fill((1, 10, 10), 13, refresh=True)
-    visual = viewer.window._qt_viewer.layer_to_visual[layer]
+    visual = qt_viewer.layer_to_visual[layer]
```
Comment (jni) [inline napari/_vispy/_tests/test_vispy_labels_layer.py:70]: Are these changes necessary? (make_napari_viewer -> manually building a qtviewer) They seem to be going backwards. Do we want a QtViewer fixture instead?

File: napari/_vispy/_tests/test_vispy_labels_layer.py [hunk 9]
```diff
@@ -95,7 +99,7 @@ def test_labels_painting_with_mouse(
         position=(0, 10, 10),
         dims_displayed=(0, 1),
     )
-    visual = viewer.window._qt_viewer.layer_to_visual[layer]
+    visual = qt_viewer.layer_to_visual[layer]
```
Comment (jni) [inline napari/_vispy/_tests/test_vispy_labels_layer.py:95]: As above — this feels like an anti-pattern — we should have a way to provide a QtViewer that doesn't involve three separate (and repeated) lines of code?

File: napari/_qt/layer_controls/qt_image_controls_base.py (hunk 5 above)
Comment (Czaki) [reply napari/_qt/layer_controls/qt_image_controls_base.py]: The `self.button_grid` is a layout that contains `self.panzoom_button` and `self.transform_button`, but the layout does not have a set parentship. So when try to detect a leaking widget, the `QApplication.topLevelWidgets()` is detecting this as a top-level widget without set parent. So it is handled by from `conftest.py`:

```
        if widget.parent() is not None:
            continue
```

This class and the second one are abstract class that we do not create instance outside the tests.

File: napari/_qt/layer_controls/qt_layer_controls_base.py (hunk 6 above)
Comment (Czaki) [reply napari/_qt/layer_controls/qt_layer_controls_base.py:165]: response above

File: napari/_vispy/_tests/test_vispy_labels_layer.py (hunk 8 above)
Comment (Czaki) [reply napari/_vispy/_tests/test_vispy_labels_layer.py:70]: > Are these changes necessary?

Yes. Our teardown procedure of `Viewer` is fragile and the main place of segfaults in our test set. I have tried to fix this in #6745, but failed. And I cannot estimate how long it may take to fix this in fixture as the problem comes from the architecture of `Viewer`

File: napari/conftest.py [hunk 10]
```diff
@@ -744,15 +745,97 @@ def my_start(self):
     )
 
 
+with contextlib.suppress(ImportError):
+    from pytestqt.qtbot import QtBot
+
+    class OwnQtBot(QtBot):
```
Comment (jni) [inline napari/conftest.py]: Can you add a docstring for this? I don't understand the need for this class — or why it's under a supress ImportError guard.

File: napari/_vispy/_tests/test_vispy_labels_layer.py (hunk 9 above)
Comment (Czaki) [reply napari/_vispy/_tests/test_vispy_labels_layer.py:95]: The antipattern is using the whole `Viewer` when just `QtViewer` is enough, as it slowdown the test and exposes the next place where something may fail. 

Having this in the fixture may be a good idea, but it is the next refactor. And I would like to spend time of proposal to SDG, so do not know when find time to do this.

File: napari/conftest.py (hunk 10 above)
Comment (Czaki) [reply napari/conftest.py]: Will add docstring 

> I don't understand the need for this class — or why it's under a supress ImportError guard.

Because of `headless` test. Same reason why all fixtures have Qt imports inside, not top-level one.

**Comment by Czaki** [comment]
I have updated doc strings. I have added fixtures for `qt_viewer` and `viewer_model`

File: napari/_vispy/_tests/test_vispy_labels_layer.py (hunk 8 above)
Comment (Czaki) [reply napari/_vispy/_tests/test_vispy_labels_layer.py:70]: done

File: napari/_vispy/_tests/test_vispy_labels_layer.py (hunk 9 above)
Comment (Czaki) [reply napari/_vispy/_tests/test_vispy_labels_layer.py:95]: done

File: napari/_qt/layer_controls/qt_image_controls_base.py (hunk 5 above)
Comment (jni) [reply napari/_qt/layer_controls/qt_image_controls_base.py]: > This class and the second one are abstract class that we do not create instance outside the tests.

ooooooh ok.

My next question is — why do we only add a parent in this circumstance rather than parenting to the layout always?

File: napari/_vispy/_tests/test_vispy_labels_layer.py (hunk 9 above)
Comment (jni) [reply napari/_vispy/_tests/test_vispy_labels_layer.py:95]: LOL Thanks! I'm still upset that pytest magic means that qt_viewer is in fact linked to the viewer_model fixture, but that's not in any way obvious from the writing of the test locally. But that's inevitable within pytest. Thanks! 🙏

File: napari/conftest.py [hunk 11]
```diff
@@ -759,15 +879,121 @@ def my_start(self):
...
+    """Because we have headless test suite we cannot use
+    @pytest.fixture(autouse=True) for fixtures that are detecting
+    leaking Qt objects.
+    So we use here  trick to detect if the `qapp` fixture is used to detect test
+    that are using Qt and need to be checked for Qt objects leaks.
+
+    It may happen that some test is using Qt but is not using `qapp` fixture.
+    This should be caught during the PullRequest review.
+    If you found such a test, please add `qapp` fixture to it.
+    As lack of this fixture may end reporting of a problem in the wrong place.
```
Comment (jni) [inline napari/conftest.py]: Thanks for this note. To clarify: qapp should be used even if qtbot is used?

File: napari/conftest.py [hunk 12]
```diff
@@ -759,15 +879,121 @@ def my_start(self):
...
+
+        if widget.__class__.__module__.startswith('qtconsole'):
+            continue
+
+        if widget.objectName() == 'handled_widget':
+            continue
+
+        if widget.__class__.__name__ == 'CanvasBackendDesktop':
+            # leaking in the tests_sys_info.py
+            continue
```
Comment (jni) [inline napari/conftest.py]: Is this a TODO?

Comment (Czaki) [reply napari/conftest.py]: It is the same note lake existing from long time in `make_napari_viewer`. There is a problem of leaking `CanvasBackendDesktop`. I have found that in my case it happens in `test_sys_info.py` 
So we do not report it as a test failure, but I left a comment to simplify the life of someone who would like to solve this, to simplify his life.

File: napari/_qt/layer_controls/qt_image_controls_base.py (hunk 5 above)
Comment (Czaki) [reply napari/_qt/layer_controls/qt_image_controls_base.py]: No, because then in subclass we will need to undock first, that will create more messy code.

File: napari/_vispy/_tests/test_vispy_labels_layer.py (hunk 9 above)
Comment (Czaki) [reply napari/_vispy/_tests/test_vispy_labels_layer.py:95]: If you want, we may use `qt_viewer.viewer` attribute. As viewer_model is stored under this public attribute.

File: napari/conftest.py (hunk 11 above)
Comment (Czaki) [reply napari/conftest.py]: Creating `QtWidget` before creating an instance of `QApplication`. Moreover, `QApplication` is required to have an event loop, that is required for signals. 

But tests do not require it directly. It is enough to be requested by fixture. 

I add comment to prevent accidental removing it in followup PR.

File: napari/conftest.py (hunk 12 above)
Comment (jni) [reply napari/conftest.py]: So you are saying, this class leaks and we can't/don't know how to fix it, so we ignore it here?

Comment (Czaki) [reply napari/conftest.py]: Yes. Since long time :disappointed:

File: napari/_qt/layer_controls/qt_layer_controls_base.py [hunk 13]
```diff
@@ -239,6 +242,7 @@ def _radio_button(
         """
         action_name = f'napari:{action_name}'
         btn = QtModeRadioButton(layer, btn_name, mode, **kwargs)
+        btn.setObjectName(f'{btn_name}_button')
```
Comment (DragaDoncila) [inline napari/_qt/layer_controls/qt_layer_controls_base.py]: What is the benefit here of setting the object name? I know it's generally good practice, but was this affecting tests before?

File: napari/_qt/qt_viewer.py [hunk 14]
```diff
@@ -1174,7 +1174,9 @@ def closeEvent(self, event):
         event : qtpy.QtCore.QCloseEvent
             Event from the Qt context.
         """
-        self.layers.close()
+        if self._layers is not None:
+            # do not create layerlist if it does not exist yet.
+            self.layers.close()
```
Comment (DragaDoncila) [inline napari/_qt/qt_viewer.py:1179]: Is this to say that before, even if a layerlist hadn't been created, by accessing `self.layers`, we were creating one just to then close it?

**Review by DragaDoncila — COMMENTED** [review]
Just left two comments asking for clarification for my own sake. Agree with @lucyleeow that we should include some of this stuff in our testing docs, and maybe even do a little "crash course in napari testing" workshop for core devs or something

File: napari/_qt/qt_viewer.py (hunk 14 above)
Comment (Czaki) [reply napari/_qt/qt_viewer.py:1179]: Yes

File: napari/_qt/layer_controls/qt_layer_controls_base.py (hunk 13 above)
Comment (Czaki) [reply napari/_qt/layer_controls/qt_layer_controls_base.py]: ops. This is my debug code, where I try to detect which widgets are leaking. 
```suggestion
```

**Comment by Czaki** [comment]
> that we should include some of this stuff in our testing docs, and maybe even do a little "crash course in napari testing" workshop for core devs or something

I hope to have time to do this in not so long time

**Comment by jni** [comment]
@Czaki I've updated the docstrings and the OwnQtBot class name to be more descriptive. Please check the docstrings especially the `pytest_runtest_setup` to make sure I haven't mangled something. Otherwise thanks again for this and this is ready to merge from me! 🙏 🚀

File: napari/conftest.py [hunk 15]
```diff
@@ -759,15 +879,139 @@ def my_start(self):
...
+    Instead, here we detect whether the `qapp` fixture is being used, detecting
+    tests that use Qt and need to be checked for Qt objects leaks.
+
+    A note to maintainers: tests *may* attempt to use Qt classes but not use
+    the `qapp` fixture. This is BAD, and may cause Qt failures to be reported
+    far away from the problematic code or test. If you find any tests
+    instantiating Qt objects but not using qapp or qtbot, please submit a PR
+    adding the qtbot fixture and adding any top-level Qt widgets with::
+
+        qtbot.addWidget(widget_instance)
```
Comment (Czaki) [inline napari/conftest.py:1003]: :heart:

File: napari/conftest.py [hunk 16]
```diff
@@ -292,6 +293,125 @@ def tmp_plugin(npe2pm_: TestPluginManager):
...
+        return self._dockConsole
+
+    def patched_dock_performance(self):
+        if self._dockPerformance is None:
+            self._dockPerformance = original_dock_performance(self)
+            qtbot.addWidget(
+                self._dockPerformance, before_close_func=hide_widget
+            )
+        return self._dockPerformance
+
```
Comment (Czaki) [inline napari/conftest.py:376]: This lack of coverage is expected as I `monkeypatch` all possible properties, that may spawn qt object. 

Should I mark it with `# pragma: no cover`?

**Review by jni — APPROVED** [review]
